packages/net/_test.pony

```diff
@@ -669,3 +673,124 @@ class _TestTCPProxyNotify is TCPConnectionNotify
 
   fun ref connect_failed(conn: TCPConnection ref) =>
     _h.fail_action("sender connect failed")
+
+class _TestTCPConnectionFailed is UnitTest
+  fun name(): String => "net/TCPConnectionFailed"
+
+  fun ref apply(h: TestHelper) =>
+    h.expect_action("connection failed")
+
+    let host = "127.0.0.1"
+    let port = "7669"
+
+    try
+      let connection = TCPConnection(
+        h.env.root as AmbientAuth,
+        object iso is TCPConnectionNotify
+          let _h: TestHelper = h
+
+          fun ref connected(conn: TCPConnection ref) =>
+            _h.log("log: connected to a non-existent host and port")
+            _h.fail("fail: connected to a non-existent host and port")
+            _h.complete(false)
+            conn.close()
+
+          fun ref connect_failed(conn: TCPConnection ref) =>
+            _h.log("log: connection failed")
+            _h.complete_action("connection failed")
+            _h.complete(true)
+        end,
+        host,
+        port)
+      h.long_test(20_000_000_000)
+      h.dispose_when_done(connection)
+    else
+      h.fail("error creating TCPConnection")
+      h.complete(false)
+    end
+
+actor _Connector
+  be connect(h: TestHelper, host: String, port: String) =>
+    try
+      let connection = TCPConnection(
+        h.env.root as AmbientAuth,
+        object iso is TCPConnectionNotify
+          let _h: TestHelper = h
+
+          fun ref connected(conn: TCPConnection ref) =>
+            _h.fail("client connected to a closed listener")
+            _h.complete(false)
+            conn.close()
+
+          fun ref connect_failed(conn: TCPConnection ref) =>
+            _h.complete_action("client connection failed")
+            _h.complete(true)
+
+          fun ref closed(conn: TCPConnection ref) =>
+            _h.log("client closed")
+        end,
+        host,
+        port)
+      h.dispose_when_done(connection)
+    else
+      h.fail("unable to create connection")
+      h.complete(false)
+    end
+
+class _TestTCPConnectionToClosedServerFailed is UnitTest
+  """
+  Check that you can't connect to a closed listener.
+  """
+  fun name(): String => "net/TCPConnectionToClosedServerFailed"
+  fun exclusion_group(): String => "network"
+
```

Comment:
This test lacks a complete(true) call. You should have that for any "success" cases.

Reply:
The `complete(true)` call is in `_Connector`, in the `connect_failed` function of the notify object.

Reply:
Ah I missed the _Connector call